src/crystal/system/unix/process.cr

```diff
@@ -354,25 +359,113 @@ struct Crystal::System::Process
 
     ::Dir.cd(chdir) if chdir
 
-    lock_write { execvpe(*prepared_args, LibC.environ) }
+    execvpe(*prepared_args, LibC.environ)
   end
 
-  private def self.execvpe(command, argv, envp)
-    {% if LibC.has_method?("execvpe") %}
-      LibC.execvpe(command, argv, envp)
+  private def self.execvpe(file, argv, envp)
+    {% if LibC.has_method?("execvpe") && !flag?("execvpe_impl") %}
+      lock_write { LibC.execvpe(file, argv, envp) }
     {% else %}
-      execvpe_impl(command, argv, envp)
+      execvpe_impl(file, argv, envp)
     {% end %}
   end
 
+  DEFAULT_PATH = "/usr/bin:/bin"
+
   # Darwin, DragonflyBSD, and FreeBSD < 14 don't have an `execvpe` function, so
   # we need to implement it ourselves.
-  # FIXME: This is a stub implementation which simply sets the environment
-  # pointer. That's the same behaviour as before, but not correct. Will fix in a
-  # follow-up.
-  private def self.execvpe_impl(command, argv, envp)
-    LibC.environ = envp
-    LibC.execvp(command, argv)
+  # This method runs between `fork` and `exec` and must be very cautious, such
+  # as no memory allocations.
+  private def self.execvpe_impl(file : String, argv : LibC::Char**, envp : LibC::Char**)
+    if file.empty?
+      Errno.value = Errno::ENOENT
+      return
+    end
+
+    # When file contains a slash, it's already a pathname that we should execute.
+    if file.includes?("/")
+      lock_write { LibC.execve(file, argv, envp) }
+
+      # Glibc implements a fallback if execve fails with ENOEXEC which tries
+      # executing `file` with `/bin/sh`. This is a legacy compatibility feature and
+      # has security concerns. We implement the behaviour of `execvpex`.
+      return
+    end
+
+    path = if path_ptr = LibC.getenv("PATH")
+             Slice.new(path_ptr, LibC.strlen(path_ptr))
+           else
+             DEFAULT_PATH.to_slice
+           end
+
+    if file.bytesize > LibC::NAME_MAX
+      Errno.value = Errno::ENAMETOOLONG
+      return
+    end
+
+    buffer = uninitialized UInt8[LibC::PATH_MAX]
+
+    io_memory_storage = uninitialized ReferenceStorage(IO::Memory)
+    io = IO::Memory.unsafe_construct(pointerof(io_memory_storage), buffer.to_slice)
```

Comment:
This fails to build 😭 
```
  In src/crystal/system/unix/process.cr:408:39
  
   408 | io_memory_storage = uninitialized ReferenceStorage(IO::Memory)
                                           ^
  Error: Can't instantiate ReferenceStorage(T) with T = IO::Memory+ (T must be a reference type)
```

This problem hasn't come up when building `process_spec.cr` on its own. It is only a problem when building the full spec suite which defines some sublcasses in https://github.com/crystal-lang/crystal/blob/7572ca25334002e14f15df4e27aa959a2f3dad08/spec/std/io/delimited_spec.cr#L24

Looks like a bug/missing feature in `ReferenceStorage`.
I guess we'll have to do without for now.